cirq/circuits/qasm_output.py

```diff
@@ -52,14 +52,20 @@ def _qasm_(self,
                args: protocols.QasmArgs) -> str:
         args.validate_version('2.0')
         return args.format(
-                'u3({0:half_turns},{1:half_turns},{2:half_turns}) {3};\n',
-                self.theta, self.phi, self.lmda, qubits[0])
+            'u3({0:half_turns},{1:half_turns},{2:half_turns}) {3};\n',
+            self.theta, self.phi, self.lmda, qubits[0])
 
     def __repr__(self) -> str:
         return 'cirq.QasmUGate({}, {}, {})'.format(self.lmda,
                                                    self.theta,
                                                    self.phi)
 
+    def _unitary_(self) -> np.ndarray:
+        rz_phi_matrix = protocols.unitary(ops.Rz(self.phi * np.pi))
+        ry_theta_matrix = protocols.unitary(ops.Ry(self.theta * np.pi))
+        rz_lambda_matrix = protocols.unitary(ops.Rz(self.lmda * np.pi))
+        return rz_phi_matrix.dot(ry_theta_matrix.dot(rz_lambda_matrix))
```

Comment:
Optional: You can also compute the four elements of this 2x2 matrix using formula (2) in [this paper](https://arxiv.org/abs/1707.03429). Something like this:

```python
pre_z_phase = np.exp(1j * np.pi * self.lmda)
post_z_phase = np.exp(1j * np.pi * self.phi)
y_phase = np.exp(1j * np.pi * np.theta)
c, s = y_phase.real, 1j * y_phase.imag
return np.array([[pre_z_phase.conj() * post_z_phase.conj() * c, ...]])
```

This should be a bit faster. The existing formulas would make a good unit test though.

Reply:
I cleaned up the code a bit more, but didn't go all the way down to writing out the matrix. (This is not a method on a hot path.) I referenced that paper as the source.